base/strings/iteration.jl

```diff
@@ -0,0 +1,120 @@
+# A specialized iterator for EachIndex of strings
+struct EachStringIndex{T<:AbstractString}
+    s::T
+end
+keys(s::AbstractString) = EachStringIndex(s)
+
+length(e::EachStringIndex) = length(e.s)
+first(::EachStringIndex) = 1
+last(e::EachStringIndex) = lastindex(e.s)
+eltype(::Type{<:EachStringIndex}) = Int
+
+# Iteration over StringNext
+#
+# Any new subtype of AbstractString, should override
+#
+#   next(::StringNext{MyString}, state)
+#
+# to provide iteration over the string and its indices. All other iteration methods,
+# including iteration over strings, iteration over pairs, indexing into string,
+# iteration over indicies alone are derived from this method.
+
+const StringNext{T<:AbstractString} = Iterators.Next{T, EachStringIndex{T}}
+StringNext(x::T) where {T<:AbstractString} = Next(x)
+StringNext(x::T, idx) where {T<:AbstractString} = Next(x, idx)
+StringNext(x::T, idx, itr) where {T<:AbstractString} = Next(x, idx, itr)
+
+start(sp::StringNext) = 1
+function done(s::StringNext, i)
+    if isa(i, Integer)
+        return i > ncodeunits(s.data)
+    else
+        throw(MethodError(done, (s, i)))
+    end
+end
+function next(s::StringNext, i)
+    if isa(i, Integer) && !isa(i, Int)
+        return next(s, Int(i))
+    else
+        throw(MethodError(next, (s, i)))
+    end
+end
+
+# Derive iteration over pairs from `StringNext`
+const StringPairs{T<:AbstractString} = Iterators.Pairs{Int, Char, EachStringIndex{T}, T}
```

Comment:
According to @Keno this is:
> the kind of Pairs iterator you get when you call `pairs` on a string of type `T`

Reply:
Can it be written something like `const StringPairs{T<:AbstractString} = typeof(pairs(T())` (I had this problem creating a const type alias for view(Vector{UInt8}), the type is really convoluted, so I ended up with `const nobytes = UInt8[]; const ByteView = typeof(view(nobytes))`

Reply:
No, I don't think that will work since there is no actual `T` at the time of definition. The RHS needs to use parametric type syntax to work – despite the appearance, this is not a normal assignment, it's what used to be written as `typealias`.